enhancements/multi-arch/multiarch-manager-operator.md

```diff
@@ -0,0 +1,836 @@
+last-updated: 2023-05-18
+tracking-link:
+  - https://issues.redhat.com/browse/MIXEDARCH-215
+see-also:
+  - "/enhancements/multi-arch/heterogeneous-architecture-clusters.md"
+---
+
+# Multiarch Manager Operator
+
+## Summary
+Recently, cloud vendors have started offering compute nodes leveraging different
+CPU architectures. With the rising use and support of multi-architecture compute
+nodes in Openshift clusters, users and cluster administrators face new
+challenges when deploying workloads. Primarily, the scheduler does not consider
+the images' compatibility with the CPU architectures when filtering nodes. We
+must assume that the images used for deploying workloads would only support some
+architecture the cluster could consist of. The typical workaround is to manually
+add affinities to the pod spec to ensure that the pods land on nodes where the
+image binaries can be executed. This strongly impacts the user experience: it
+couples the set of compatible architectures at a given point in time to the pod
+spec, does not scale well, and is challenging to maintain. The
+`multiarch-manager-operator` aims to automate the inspection of the container
+images, derive a set of architectures supported by a pod and use it to
+automatically define strong predicates based on the `kubernetes.io/arch`
+label in the pod's nodeAffinity.
+
+## Motivation
+- Easy adoption of multi-architecture compute clusters by customers with varying
+  workload images which might not be built for specific architectures
+- Provides users a "clean" user experience without having to worry about
+  manually modifying deployment specs when deploying workloads on a multi-arch
+  compute clusters
+
+### User Stories
+- As a user who wants to deploy workloads on a multi-arch compute cluster, I
+  want the pods to be scheduled on the appropriate subset of nodes based on the
+  architectures supported by their containers' images so that the coupling
+  between the characteristics of the images and the nodeAffinity declared in
+  the pod spec template does not require future manual maintenance due to image
+  updates provided by the external stake-holders maintaining them
+- As a cluster administrator, when I add a new node of a secondary architecture
+  to a cluster, I do not want to taint the nodes to be unschedulable to ensure
+  that workloads continue to deploy and run as expected given that the images
+  used for deploying workloads may not support the above said secondary
+  architecture
+- As a cluster administrator who wants to migrate from a single-arch to a
+  multi-arch compute cluster, I do not want to rebuild each container image for
+  all the architectures that are present in the cluster
+- As a user who wants to deploy workloads on a multi-arch compute cluster, I do
+  not want to modify every pod (or any resources owning pods) spec to include a
+  node affinity predicate to filter the nodes at scheduling time, based on the
+  compatible architectures
```

Comment:
Any node selector/affinity about the node architecture manually set by the users is prioritized: the controller will not overwrite those predicates externally set by the users.

The requests/limits concerns are interesting and worth a broader discussion IMO. However, I believe they are out of the scope of this EP.

Reply:
This is a good point and needs to be covered in documentation letting the user know that their pod can be scheduled on nodes with any of the architectures matching the images. This would then require manual intervention from the user to either exclude the namespaces where such pods are deployed or specify their own node selector arch preference.

Reply:
added this concern to the risks and mitigations section